### PR TITLE
Display filters in search summary in order of application

### PR DESCRIPTION
**Pull request by jazairi**
#### Why these changes are being introduced:

The search summary currently displays filters in alphabetical order by key. This is confusing, as one would expect them to show up in the order they were applied.

#### Relevant ticket(s):

* [GDT-224](https://mitlibraries.atlassian.net/browse/GDT-224)

#### How this addresses that need:

Many (if not all) Rails helpers pertaining to URLs (`to_query`, `to_params`, routing helpers, etc.) sort their inputs lexicographically. This means that we cannot use any of these methods, or the `params` object, to pass query params into the Enhancer, as they will not remain ordered and will instead be alphabetized everywhere in the application.

This commit seeks to fix this problem by passing
`request.original_url` into the Enhancer and building filter URLs manually. It's messier than using the Rails convenience methods, particularly because we're now mostly working with strings rather than hashes, but it felt less complex than the other approach considered of storing filter order in a cookie.

Prior to this change, both the Enhancer and QueryBuilder iterated over arrays of supported params and added them to the query if they showed up in the `params` object. This is now reversed: the extraction methods in both models now iterate over the URL params and confirm that they exist in the list of supported params before adding them to the query. The reason for this change is to ensure that filters (and other params) are added to the query in the order in which they appear in the original URL, not the order in which they are listed in the array of supported params.

#### Side effects of this change:

* Fixes a bug in the filter helper that prevents single-valued filters from being applied. Currently, the only single-valued filter we have is `literaryFormFilter`, which is not used anywhere, but it makes sense to support all fields that TIMDEX API provides.
* Updates several tests to accommodate the refactored Enhancer and QueryBuilder.
* Moves the constants that recur in the Enhancer and QueryBuilder into a new module, QueryElements.
* Moves all of the filter helper methods related to adding/removing filters next to each other for better readability.
* Regenerates several cassettes. (It's possible that not all of these needed to be regenerated; I erred on the side of more.)

##### Accessibility

- [x] ANDI or WAVE has been run in accordance to [our guide](https://mitlibraries.github.io/guides/basics/a11y.html).
- [ ] This PR contains no changes to the view layer.
- [ ] New issues flagged by ANDI or WAVE have been resolved.
- [ ] New issues flagged by ANDI or WAVE have been ticketed (link in the Pull Request details above).
- [x] No new accessibility issues have been flagged.

##### New ENV

- [ ] All new ENV is documented in README.
- [ ] All new ENV has been added to Heroku Pipeline, Staging and Prod.
- [x] ENV has not changed.

##### Approval beyond code review

- [ ] UXWS/stakeholder approval has been confirmed.
- [x] UXWS/stakeholder review will be completed retroactively.
- [ ] UXWS/stakeholder review is not needed.

##### Additional context needed to review

In terms of confirming functionality: ensure that filters from _every_ category (including `literaryFormFilter`) can be added and removed successfully, and appear in the search summary panel in the order you add them.

In terms of the code: one thing I am particularly interested in is where this is adequately documented. Please let me know if (and where) you'd like to see additional explanation.

##### Code

- [ ] I have confirmed that the code works as intended.
- [x] Any CodeClimate issues have been fixed or confirmed as
added technical debt.

##### Documentation

- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message).
- [ ] The documentation has been updated or is unnecessary.
- [ ] New dependencies are appropriate or there were no changes.

##### Testing

- [ ] There are appropriate tests covering any new functionality.
- [ ] No additional test coverage is required.

[GDT-224]: https://mitlibraries.atlassian.net/browse/GDT-224?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/helpers/filter_helper.rb [hunk 1]
```diff
@@ -12,48 +12,29 @@ def add_filter(query, filter, term)
...
-  end
-
-  def gdt_sources(value, category)
-    return value if category != :sourceFilter
-
-    return 'Non-MIT institutions' if value == 'opengeometadata gis resources'
-
-    return 'MIT' if value == 'mit gis resources'
-
-    value
```
Comment (jazairi) [inline app/helpers/filter_helper.rb:43]: Note to reviewer: this is the result of reordering the methods in the filter helper, as noted in the side effects. (See also lines 70-123.)

File: app/helpers/filter_helper.rb [hunk 2]
```diff
@@ -78,8 +59,66 @@ def applied_filters(query)
...
+    return 'Non-MIT institutions' if value == 'opengeometadata gis resources'
+
+    return 'MIT' if value == 'mit gis resources'
+
+    value
+  end
+
+  private
+
+  # This kind of manual URL construction is required because any core exntesions method involving params, including
```
Comment (JPrevost) [inline app/helpers/filter_helper.rb:97]: Typo: `exntesions`

**Comment by jazairi** [comment]
I noted this in JIRA, but there's a bug in this PR that still allows filters to be displayed out of order.

**Comment by jazairi** [comment]
I noted this in JIRA, but there's a bug in this PR that still allows filters to be displayed out of order.

In the query object, multiple filters in the same category are assigned to the same key, since keys are unique in Ruby hashes.

This means that when we apply the filters in the example below (subject: transportation, then place: massachusetts, then subject: maps), the query is built like so:

```
{
  subjectsFilter: ['transportation', 'maps'],
  locationsFilter: ['massachusetts']
}
```

And when we iterate through that hash, the filters get inevitably grouped by key:

```
[{ subjectsFilter: 'transportation', subjectsFilter: 'maps', locationsFilter: 'massachusetts' }]
```

I think we need to rework our approach to this problem. It's likely a good use case for cookies, as we've discussed elsewhere.

**Pull request closed without merging**